### PR TITLE
Add interferometric likelihood and linear solver

### PR DESCRIPTION
**Pull request by nanz6**
Here is an updated version of PR #400, for the interferometric image fitting.
It is based on the newest version of Lenstronomy and I have packaged the log_likelihood and linear_solve functions for interferometric images into new functions that will be invoked by the general log_likelihood and linear_solve functions if 'likelihood_method' is assigned to 'interferometry_natwt'.

With this pull request merged, lenstronomy is able to do the parametric fitting for interferometric images.

Changes speciafically:

ImageData:

Variables added: log_likelihood_constant and likelihood_method
i. log_likelihood_constant allows user to input a float that will be added to the output of the log_likelihood function.
ii. likelihood_method allows user to choose the likelihood method, the default is 'diagonal', which will compute the likelihood in the normal way of lenstronomy. If the user set this parameter to 'interferometry_natwt', it will compute the logL using our natwt special method.

New functions:
i. log_likelihood_interferometry(self, model): returns the logL of interferometric method. There are math formulae in the comments of this function, explains how it is working.
ii. likelihood_method: passes 'likelihood_method' to ImageLinearFit to decide which linear solver method to use.

Raise error if likelihood_method is not input as 'diagonal' or 'interferometry_natwt'.

ImageLinearFit:

Add _image_linear_solve_natwt_special,  _image_linear_solve_interferometry_natwt functions to execute the linear solving for the interferometric method. There are math equations in the comments explaining the method.

Changes on Test files:

test_imaging_data.py:
Added one more raiseError checking. If the likelihood_method is not input as 'diagonal' or 'interferometry_natwt', it should raise a ValueError.

Add and fix the test files:
test_imaging_data_with_interferometric_changes.py
test_image_linear_solve_with_interferometric_changes.py
test_image_model_with_interferometric_changes.py
to test the interferometry natwt likelihood and linear solver changes.

### CODE REVIEW

**Review by sibirrer — CHANGES_REQUESTED** [review]
This looks great, thank you very much! I have some minor comments about reference in interferometry but otherwise looks great!
I make it as 'request changes' but once these are done, ready to be merged

File: lenstronomy/Data/imaging_data.py [hunk 1]
```diff
@@ -100,8 +113,44 @@ def log_likelihood(self, model, mask, additional_error_map=0):
...
+        if self._logL_method == 'interferometry_natwt':
+            return self.log_likelihood_interferometry(model)
+        
         C_D = self.C_D_model(model)
         X2 = (model - self._data) ** 2 / (C_D + np.abs(additional_error_map)) * mask
         X2 = np.array(X2)
         logL = - np.sum(X2) / 2
         return logL
+    
+    def log_likelihood_interferometry(self, model):
```
[the remark this answers is not in the document]
Comment (nanz6) [reply lenstronomy/Data/imaging_data.py:126]: For the likelihood function, I should refer my paper there, but sadly it is not finished. So I cite one paper talking about the crucial step giving the relation of the covariance matrix and PSF for the natural weighting interferometry images.
I will change that after my paper is finished.

**Review by sibirrer — APPROVED** [review]
sounds good! I merge it and you can update the reference later. Thank you very much!

**Comment by sibirrer** [comment]
@nanz6, I just noticed that the image_model.py routines were not updated. Currently the features only work when you solve the linear optimization. Is it possible to also make the function work with the ImageModel() class?
Thanks, Simon

**Comment by nanz6** [comment]
Hi @sibirrer, you are correct. The current features allow the fitting of the interferometric images, that's why I changed just the imageing data, where the likelihood function is defined, and linear_fit. 
I am not very sure what you are referring to for ImageModel(). Is that for image simulation?

**Comment by sibirrer** [comment]
> I am not very sure what you are referring to for ImageModel(). Is that for image simulation?
This is the class here: https://github.com/lenstronomy/lenstronomy/blob/main/lenstronomy/ImSim/image_model.py
This is effectively the base class for image simulation. The LinearFit() class inherits these features and adds the linear solver functionality. If you want to e.g. simulate an image without the linear solver for a given amplitude parameters, this is the class that you want to use.

**Comment by sibirrer** [comment]
> I am not very sure what you are referring to for ImageModel(). Is that for image simulation?
This is the class here: https://github.com/lenstronomy/lenstronomy/blob/main/lenstronomy/ImSim/image_model.py
This is effectively the base class for image simulation. The LinearFit() class inherits these features and adds the linear solver functionality. If you want to e.g. simulate an image without the linear solver for a given amplitude parameters, this is the class that you want to use.

**Comment by nanz6** [comment]
> > This is the class here: https://github.com/lenstronomy/lenstronomy/blob/main/lenstronomy/ImSim/image_model.py
> > This is effectively the base class for image simulation. The LinearFit() class inherits these features and adds the linear solver functionality. If you want to e.g. simulate an image without the linear solver for a given amplitude parameters, this is the class that you want to use.

I think making the images is the same for the interferometric case, if you input all the parameters, specify amplitudes and give the pixelbased PSF, you will get the desired convolved model image. In the past PRs, we have updated the primary beam and the option to diable the PSF auto-normalization, which is helpful for making interferometric images and is part of the update to the image model.

By the way, the difficulty of making a fake interferometric data image is to make the noises, which is usually done by the data calibration package, CASA, so I don't prepare to make that for lenstronomy. (Although I don't think you are referring to this thing. )

So I am not sure what you think we should update to the image model. 

There is surely a conflict with the Model Plot for the interferometric data class if we choose linear_solver=True, (things are good if linear_solver=False and we give correct amplitudes). I will fix that in the future.

**Comment by sibirrer** [comment]
@nanz6 ah I see there is already the primary beam in this one. I just did not look at what was already there. never mind.
I also see the issue on the model plot. It hopefully does not take too much to fix it, otherwise let me know! Thank you very much!

**Pull request merged**